### PR TITLE
[air/execution] Actor+future manager 1: Abstractions

### PR DESCRIPTION
**Pull request by krfricke**
Signed-off-by: Kai Fricke <kai@anyscale.com>

## Why are these changes needed?

This PR adds the abstractions for an event-based manager around Ray core primitives.

The RayEventManager will be used by the AIR libraries to start and stop actors and to schedule futures on them.
Instead of imperatively resolving the futures manually using `ray.wait` and `ray.get`, we define callbacks that are invoked when specific event happen. In particular, we will be able to specify callbacks for the following events:

- An actor started
- An actor terminated (gracefully_
- An actor failed
- An actor task resolved successfully
- An actor task failed

This event processing matches the patterns we currently implement in our libraries. By abstracting this into a separate management class, we remove the Ray core primitive lifecycle management from our library code. This leads to better testability and maintainability of our code base, and it will make it easier for us to extend our functionality in the future.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/air/execution/actor_manager.py
Comment (gjoliver) [inline python/ray/air/execution/actor_manager.py:220]: why cancel the unfinished tasks?
the contract current rllib actor manager has with the rest of the codebase is return everything that succeeded, return error for anything that failed.
also, does FutureResult contain the actual data or just the object_ref?

File: python/ray/air/execution/actor_manager.py
Comment (gjoliver) [inline python/ray/air/execution/actor_manager.py:205]: any ways to specify timeout?

File: python/ray/air/execution/actor_manager.py
Comment (gjoliver) [inline python/ray/air/execution/actor_manager.py:178]: do we have to batch future results ourselves?
it's usually a lot more efficient to handle multiple futures together.

**Review by gjoliver — COMMENTED** [review]
some high level questions comparing this with the current actor manager RLlib uses.

File: python/ray/air/execution/actor_spec.py
Comment (krfricke) [reply python/ray/air/execution/actor_spec.py:69]: I see. That makes sense - I think we can just use an actor ID then and treat actor spec as ephemeral

File: python/ray/air/execution/event.py
Comment (krfricke) [reply python/ray/air/execution/event.py:21]: We do need to identify the actor (e.g. via ID) but we can probably get rid of the specs. The caller can track the specs itself if needed (we'll likely do that for Tune to enable actor re-use)

File: python/ray/air/execution/actor_manager.py
Comment (krfricke) [reply python/ray/air/execution/actor_manager.py:220]: FutureResult should contain the actual data (I can make this more clear in the docs).

The reason we cancel the remaining tasks is because these are _synchronous_ futures. E.g. in Ray Train if we schedule training of episode X on N actors, we only care about the results if we can get all N results.

Can you elaborate more on how rllib does this? Are these actually synchronous futures that should all return at the same time? My understanding was that these are unsynchronous in which case the behavior is as described.

File: python/ray/air/execution/actor_manager.py
Comment (krfricke) [reply python/ray/air/execution/actor_manager.py:205]: We can add this. Is rllib using this already or is that future use?

File: python/ray/air/execution/actor_manager.py
Comment (krfricke) [reply python/ray/air/execution/actor_manager.py:178]: Is this only about the results (outputs) or also about scheduling (input)?

We could emit multiple results if the futures happen to resolve at the same time. For Tune that would actually complicate things a bit in processing as we do everything one-by-one and no batched processing.

For Train, where we actually do batched processing, we would do this with sync futures, i.e. futures that we require to resolve at the same time.

Maybe you can share a bit more how exactly rllib would like to interact with this. I.e. how are futures scheduled, how are they expected to resolve, and how are they processed.

File: python/ray/air/execution/actor_manager.py
Comment (krfricke) [reply python/ray/air/execution/actor_manager.py:204]: I'm holding back on adding this right now until we clarified the use

File: python/ray/air/execution/actor_spec.py
Comment (krfricke) [reply python/ray/air/execution/actor_spec.py]: Updated. Also renamed `ActorInfo` to `TrackedActor`.

File: python/ray/air/execution/event.py
Comment (krfricke) [reply python/ray/air/execution/event.py]: Updated.

File: python/ray/air/execution/tracked_actor.py
Comment (amogkam) [inline python/ray/air/execution/tracked_actor.py:30]: May want to rename these methods to set_*, to make explicit that these are only setting the callbacks, but are not actually executing it.

Speaking of which, is there a separate API to actually trigger the callbacks?

File: python/ray/air/execution/event_manager.py
Comment (amogkam) [inline python/ray/air/execution/event_manager.py:83]:
```suggestion
        """Yield control to event manager to await events and invoke callbacks.
```

**Review by amogkam — COMMENTED** [review]
Thanks @krfricke! One question I have is on the callback interface.

How would you specify logic like "if any actor errors, then terminate all actors"? It doesn't seem possible with the current API since the callback only takes in a single actor as input: `Callable[[TrackedActor], None]`.

Should the callbacks be specified at the manager level and not at the individual actor/task level?

File: python/ray/air/execution/tracked_actor.py
Comment (krfricke) [reply python/ray/air/execution/tracked_actor.py:30]: IMO the `on_` API is pretty clear about not executing the callbacks: `future.on_error(handle_error).on_timeout(handle_timeout)` etc - why would anyone assume the callbacks are executed immediately?

The callbacks will be triggered whenever the event arrives. E.g. when the future resolves, the `on_result` callback will be called. When the future errors, the `on_error` callback will be called.

For this we have to yield control to the event manager via `manager.wait()`. The alternative would be to do this in a separate thread, but then we run into problems with thread-safety, so I think for now that's the best way to go forward.

**Comment by krfricke** [comment]
To your question, the API could look something like this:

```
manager = RayEventManager()

def handle_actor_error(tracked_actor, exception):
    for actor in manager.all_actors:
        manager.remove_actor(actor)

actor = manager.add_actor(cls=ActorClass).on_error(handle_actor_error)
```

File: python/ray/air/execution/tracked_actor.py
Comment (matthewdeng) [inline python/ray/air/execution/tracked_actor.py:24]: Should these be made part of the constructor instead? I feel like mutating these with `on_X` feels a bit strange? Though let me know if this is a standard pattern that I'm just unaware of.

File: python/ray/air/execution/tracked_actor.py
Comment (matthewdeng) [inline python/ray/air/execution/tracked_actor.py:13]: Can you document the expected way for a user to create one of these, via `RayEventManager.add_actor()`?

File: python/ray/air/execution/event_manager.py
Comment (matthewdeng) [inline python/ray/air/execution/event_manager.py:99]: Reading through this I now see why this is included here... One thing I'm wondering if we should expose a separate event type for `RESOURCES` scheduling as well, or if this can be encapsulated by Actor/Task errors entirely.

File: python/ray/air/execution/barrier.py
Comment (matthewdeng) [inline python/ray/air/execution/barrier.py:46]: nit: Should `*data` instead just be `data: Any`? To ensure compatibility with `get_errors` without wrapping/unwrapping.

File: python/ray/air/execution/barrier.py
Comment (matthewdeng) [inline python/ray/air/execution/barrier.py:60]: I'm not sure about the different use-cases for this, but one question I had while reading the API is if it would be valuable to split this `on_completion` callback into two, with the second being a one specfically for `on_first_error`?

**Review by matthewdeng — COMMENTED** [review]
Thanks for putting this together, overall a fan of how clean/simple these APIs feel!

File: python/ray/air/execution/event_manager.py
Comment (krfricke) [reply python/ray/air/execution/event_manager.py:251]: Only tracked in the event manager. Tracked actors should be easy to pass around, so we shouldn't attach any state to them

File: python/ray/air/execution/tracked_actor.py
Comment (krfricke) [reply python/ray/air/execution/tracked_actor.py:24]: It's the standard pattern in asynchronous programming, e.g. javascript/jquery.

Main reason we want to use a method and not a constructor is that the developer actually never constructs these, they get the instances from the event manager and then can specify callbacks.

File: python/ray/air/execution/event_manager.py
Comment (krfricke) [reply python/ray/air/execution/event_manager.py:99]: resources is basically == actor start. In the current implementation it works well, so unless we identify the need to do this separately, let's keep it together. Adding that would be a simple change but I don't think it's needed.

File: python/ray/air/execution/barrier.py
Comment (krfricke) [reply python/ray/air/execution/barrier.py:46]: This is to make barrier compatible with a host of different callback-invoking functions that have different numbers of arguments. E.g. `tracked_task.on_error` will call `barrier.error(exception)` while `tracked_actor_task.on_error` will call `barrier.error(tracked_actor, exception)`.

Unwrapping is up to the caller who controls which kind of callbacks invoke the `barrier.error`

File: python/ray/air/execution/barrier.py
Comment (krfricke) [reply python/ray/air/execution/barrier.py:60]: I think this does make sense, I'll update the abstraction.

File: python/ray/air/execution/tracked_actor.py
Comment (krfricke) [reply python/ray/air/execution/tracked_actor.py:13]: I can add an example

Comment (krfricke) [reply python/ray/air/execution/tracked_actor.py:13]: Added

File: python/ray/air/execution/barrier.py
Comment (krfricke) [reply python/ray/air/execution/barrier.py:58]: Updated

**Comment by krfricke** [comment]
Thanks @matthewdeng, replied to comments and updated PR, ptal

**Review by gjoliver — COMMENTED** [review]
This looks quite good to me.
But @krfricke, we had a long discussion with @scv119 about which part of this PR should hopefully go into Core, and which part may have compatibility issues with Core's roadmap.
Can we wait for comments from @rkooo567 before we merge this?
This will give Core team a chance to 1) get a good understanding of what we want 2) foresee any future issues.
Thanks.

File: python/ray/air/execution/barrier.py
Comment (scv119) [inline python/ray/air/execution/barrier.py:6]: let's remove DeveloperAPI? It implies something Ray users could potentially use.

**Comment by scv119** [comment]
Similar to the resource manager PR, we can use opportunity to improve Ray core API as well, can we also write a requirement doc?

**Comment by rkooo567** [comment]
+1 in the requirement (but the PR doesn't need to be blocked by us) + not annotating with DeveloperAPI for now. It'd be great for us to understand the pain point to find the opportunity to improve core APIs. 

Also a couple of questions. 

1. Looks like it is a wrapper to allow callback-based programming for Ray APIs. Why do you guys prefer this way over "regular way of using ray APis"? 
2. Also curious existing APIs like `concurrent.future` APIs don't work? https://docs.ray.io/en/latest/ray-core/actors/async_api.html#objectrefs-as-concurrent-futures-futures. Looks like tracker* classes are very similar to concurrent.future (or is it a wrapper on top of it?).
3. Also what's the reason why we build another Pool API instead of using/improving existing ones (e.g., ActorPool)? Is it like the use case is very niche, or the existing API is not sufficient?

**Comment by amogkam** [comment]
3. Hmm I don’t think this is really another Pool API. It’s not submitting tasks to some queue where any actor can pull from and execute. We need the ability to schedule tasks on specific actors. It’s more like List[Actor] essentially.

**Comment by gjoliver** [comment]
> +1 in the requirement (but the PR doesn't need to be blocked by us) + not annotating with DeveloperAPI for now. It'd be great for us to understand the pain point to find the opportunity to improve core APIs.
> 
> Also a couple of questions.
> 
> 1. Looks like it is a wrapper to allow callback-based programming for Ray APIs. Why do you guys prefer this way over "regular way of using ray APis"?
> 2. Also curious existing APIs like `concurrent.future` APIs don't work? https://docs.ray.io/en/latest/ray-core/actors/async_api.html#objectrefs-as-concurrent-futures-futures. Looks like tracker* classes are very similar to concurrent.future (or is it a wrapper on top of it?).
> 3. Also what's the reason why we build another Pool API instead of using/improving existing ones (e.g., ActorPool)? Is it like the use case is very niche, or the existing API is not sufficient?

For 1. it actually makes it much easier to build our libraries. happy to chat more about it.
For 2. Kai has a lot of thoughts about this that he can share.
And for 3. I don't think this is an actor pool either. It does group a set of actors together, but these actors can be completely different, and you can call / use them in  different ways. Like we are not treating them as a homogeneous set. We just want their life-cycles to be managed, so we don't have to try-catch ray.get() everywhere.

**Review by gjoliver — APPROVED** [review]
approving now. but let's make sure we address Chen and Sang's concerns / questions.

**Comment by krfricke** [comment]
> Similar to the resource manager PR, we can use opportunity to improve Ray core API as well, can we also write a requirement doc?

Yes - I'll send you the doc that motivates this PR later today.

> +1 in the requirement (but the PR doesn't need to be blocked by us) + not annotating with DeveloperAPI for now. It'd be great for us to understand the pain point to find the opportunity to improve core APIs.

Sure, I'll remove the annotations and move this into a private subpackage.

Thanks for not blocking, this is what I hoped by implementing this for Ray AIR first. This is basically dogfooding the requirements and future changes to any core-provided abstraction will still be simple.

I think the main confusion comes from the scope of the DeveloperAPI annotation - the manager is actually very useful if someone want to develop their library on top of Ray (which is what we do), but not necessarily for someone who just wants to build their own trainers.

> Also a couple of questions.
> 
>     1. Looks like it is a wrapper to allow callback-based programming for Ray APIs. Why do you guys prefer this way over "regular way of using ray APis"?
> 

It's just what we implicitly did before. In Ray Tune, we resolve training futures, if they fail we call a `_on_trial_fail` method, and if they resolve we call a `_on_trial_result` method. 
However, we had to do this in many places.
Abstracting the interaction with core away enables us to focus on library logic and not mix high-level lbirary concepts with lower level Ray primitive handling.

>     2. Also curious existing APIs like `concurrent.future` APIs don't work? https://docs.ray.io/en/latest/ray-core/actors/async_api.html#objectrefs-as-concurrent-futures-futures. Looks like tracker* classes are very similar to concurrent.future (or is it a wrapper on top of it?).

For my understanding the `concurrent.future` APIs are mostly useful to mix with other asyncio concepts, which we don't do here. It's not necessarily about asynchronicity, but more about an easier logic expression.

> 
>     3. Also what's the reason why we build another Pool API instead of using/improving existing ones (e.g., ActorPool)? Is it like the use case is very niche, or the existing API is not sufficient?

As Jun and Amog mentioned, the main difference is that we deal with a stateful and heterogeneous set of actors. I.e. we have to specifically target certain actors for certain futures, which is out of scope for the actor pool.

File: python/ray/air/execution/tracked_actor.py
Comment (krfricke) [reply python/ray/air/execution/tracked_actor.py]: The main alternative is to pass the callbacks when scheduling the objects, e.g.

```
event_manager.schedule_task(remote_fn, on_result=callback, on_error=callback)

# vs.

event_manager.schedule_task(remote_fn).on_result(callback).on_error(callback)
```

the main reason why we should keep the current proposal is that we can change callbacks later. E.g. if we gang schedule 10 training futures and one errors, we can unset the error callbacks from the other futures so we only handle one failure, etc.

File: python/ray/air/execution/barrier.py
Comment (matthewdeng) [reply python/ray/air/execution/barrier.py]: In this example you mentioned, if the following are called:

```
barrier.error(exception)
barrier.error(tracked_actor, exception)
```

What would `barrier.get_errors()` return?

Comment (krfricke) [reply python/ray/air/execution/barrier.py]: It would return

```
[[exception], [tracked_actor, exception]]
```

but you should probably not mix different caller types in one barrier.

Comment (krfricke) [reply python/ray/air/execution/barrier.py]: So what I meant is more that we don't want to have different Barrier classes for etc TrackedTask and TrackedActorTask.

File: python/ray/air/execution/tracked_actor.py
Comment (matthewdeng) [reply python/ray/air/execution/tracked_actor.py]: Right, my main concern here is around the guarantees and expectations around whether or not callbacks will be invoked, which I think is particularly difficult yet important for us to define clearly for the non-happy paths (race conditions!) in asynchronous programming.

In a simple scenario, with the current API what is the expected behavior if this sequence happens?

1. `tracked_task = event_manager.schedule_task(remote_fn)`.
2. An error occurs.
3. The caller calls `tracked_task.on_error(callback)`.

Is it guaranteed that the callback will be executed? Does it depend on whether the event manager catches/processes the error before or after (3)?

> if we gang schedule 10 training futures and one errors, we can unset the error callbacks from the other futures so we only handle one failure, etc.

Similarly, for this example would the expectation be that only one failure will ever need to be processed? Or is this meant to be an optimization to reduce additional calls? I feel that regardless we would still need to have handling in the consumer (e.g. in the `Barrier`) to reject/ignore the other errors if they happen before this logic is invoked.

To make this concise:
1. Should callbacks be static or dynamic?
2. If callbacks are dynamic, what are the guarantees around which callback(s) is invoked?

Personally I would prefer to start with statically defined callbacks, and if/when we do want to support dynamic callbacks, I'd really want for us to nail down the semantics before implementation.

**Comment by matthewdeng** [comment]
Can you make sure to add a PR description? 😄

File: python/ray/air/execution/tracked_actor.py
Comment (krfricke) [reply python/ray/air/execution/tracked_actor.py]: In your concrete example, if you call

```
tracked_task.on_error(callback)
event_manager.wait()
```

the callback will be invoked.

If you call 

```
event_manager.wait()
tracked_task.on_error(callback)
```

it won't

File: python/ray/air/execution/barrier.py
Comment (matthewdeng) [reply python/ray/air/execution/barrier.py]: Ah in that case would it make sense for `get_errors` to return a `List[List[Any]]`? I.e. how do we make sure the usage from `barrier.error()` -> `barrier.get_errors()` is intuitive?

Comment (krfricke) [reply python/ray/air/execution/barrier.py]: Ah yeah, not using `List[List[Any]]` was an oversight. Will update the PR

Comment (krfricke) [reply python/ray/air/execution/barrier.py]: I guess it's actually `List[Tuple[Any]]`. Updated

**Comment by krfricke** [comment]
After syncing with @matthewdeng offline, the final changes:

- Removed TrackedTask as we don't use this pattern in AIR at the moment. It's trivial to add back if we need it.
- Added a note in the RayEventManager docstring to make it very clear that control has to be yielded to the event manager and callbacks will only be invoked when we are in a `RayEventManager.wait()` call
- Added notes to the `Tracked*` classes that they are not to be constructed manually

**Review by matthewdeng — APPROVED** [review]

**Pull request merged**
